### PR TITLE
refactor: refactor the makefile

### PR DESCRIPTION
**Pull request by fengzixu**
Signed-off-by: fengzixu <hnustphoenix@gmail.com>

### Ⅰ. Describe what this PR did

1. simplify the content of makefile
2. supports two options about build components "docker" and "local"

### Ⅱ. Does this pull request fix one issue?

fixed:  https://github.com/dragonflyoss/Dragonfly/issues/330

#### For Linux

I run commands as below:

```
make clean
make build
make build USE_DOCKER=0  // or export USE_DOCKER=0
make clean
make build USE_DOCKER=1 //  or export USE_DOCKER=1
make install
make unisntall
make check
make uint-test
```

And then, I run scripts in the dir called hack:

```
./build-client.sh                    // build clients in the local environment.
export USE_DOCKER=1
./build-client.sh                      // build clients with docker.
./build-client.sh dfdaemon           // only build dfdaemon with docker
./build-client.sh dfget                // only build dfget with docker
export USE_DOCKER=0         // or unset USE_DOCKER
./build-client.sh dfdaemon         // only build dfdaemon in the local environment
./build-client.sh dfget                // only build dfget in the local environment
export USE_DOCKER=0   
./build-supernode                   // build supernode in the local environment.
export USE_DOCKER=01
./build-supernode                   // build supernode with docker.
``` 

#### For Mac

I run commands as below:

```
make clean
make build
make build USE_DOCKER=0  // or export USE_DOCKER=0
make clean
make build USE_DOCKER=1 //  or export USE_DOCKER=1
make install
make unisntall
make check
make uint-test
```
And then, I run scripts in the dir called hack:

```
./build-client.sh                    // build clients in the local environment.
export USE_DOCKER=1
./build-client.sh                      // build clients with docker.
./build-client.sh dfdaemon           // only build dfdaemon with docker
./build-client.sh dfget                // only build dfget with docker
export USE_DOCKER=0         // or unset USE_DOCKER
./build-client.sh dfdaemon         // only build dfdaemon in the local environment
./build-client.sh dfget                // only build dfget in the local environment
export USE_DOCKER=0   
./build-supernode                   // build supernode in the local environment.
export USE_DOCKER=1
./build-supernode                   // build supernode with docker.
```

### CODE REVIEW

**Comment by fengzixu** [comment]
@lowzj @allencloud

File: Makefile [hunk 1]
```diff
@@ -12,84 +12,24 @@
...
-
-LDFLAGS_DFGET = -ldflags "-X ${PKG}/cmd/dfget/app.Build=${BUILD}"
-LDFLAGS_DFDAEMON = -ldflags "-X ${PKG}/cmd/dfdaemon/app.Build=${BUILD}"
-
-ifeq ($(GOOS),darwin)
-    BUILD_PATH := release
-endif
+# You can assign 1 to USE_DOCKER so that you can build components of Dragonfly
+# with docker.
+USE_DOCKER=0 # Default: build components in the local environment.
```
Comment (lowzj) [inline Makefile]: How about making the env variable as the default value?

```
USE_DOCKER?=0
```

in `hack/build-client.sh`

```bash
USE_DOCKER=${USE_DOCKER:-"0"}
```

then, we can execute build commands:

```bash
# use docker
export USE_DOCKER=1
make build # or: make build USE_DOCKER=1
./hack/build-client.sh
./hack/build-client.sh dfget

# use local env
unset USE_DOCKER
make build # or: make build USE_DOCKER=0
./hack/build-client.sh
./hack/build-client.sh dfget
./hack/build-client.sh dfdaemon
```

**Comment by allencloud** [comment]
Thanks for your work. @fengzixu 
I will try to test this on my local machine.

File: Makefile [hunk 2]
```diff
@@ -101,14 +41,13 @@ uninstall:
 	./hack/install-client.sh uninstall
 .PHONY: uninstall 
 
-build-client: check-client build-dfget build-dfdaemon
```
[the remark this answers is not in the document]
Comment (fengzixu) [reply Makefile]: According to suggestions from @allencloud , we should maintain a simple makefile. `make build` and `make install` are essential for developers. If someone wants to process other customed commands, you can use scrips stored in the dir called hack.

File: hack/build-supernode.sh [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+        "0")
+            echo "Begin to build supernode in the local environment."
+            build-supernode-local
+        ;;
+        "1")
+            echo "Begin to build supernode with docker."
+            build-supernode-docker
+        ;;
+        *)
+            echo "Begin to build supernode in the local environment."
```
[the remark this answers is not in the document]
Comment (fengzixu) [reply hack/build-supernode.sh]: Done

File: Makefile (hunk 1 above)
Comment (fengzixu) [reply Makefile]: Done

**Comment by fengzixu** [comment]
@lowzj @Starnop  Request review again.

File: Makefile (hunk 2 above)
[the remark this answers is not in the document]
Comment (lowzj) [reply Makefile]: There is a balance between simple and useful. 
In daily development, we just need to build client or supernode to get an executable file at most of the time. It requires less, costs less time and use much more frequent to build one of client and supernode than both of them.
I think that the makefile should build what developers want ASAP, output enough and no extra needless information. Personally, I prefer to use scripts in `hack` if `build-client` and `build-supernode` are removed from makefile. And retaining them not only doesn't increase the makefile's complexity, but also make it more convenient and useful. WDYT?

[the remark this answers is not in the document]
Comment (fengzixu) [reply Makefile]: Done

**Comment by fengzixu** [comment]
@lowzj Ping :)

**Pull request merged**